### PR TITLE
Add Top Level Container for Saturation Function Consistency Checks

### PR DESCRIPTION
**Pull request by bska**
This PR introduces a new top-level "manager" for all saturation function consistency checks.  This component associates collections of saturation function curves with per-cell or per-region end-point definitions and provides an inteface to run all checks for all interior entities (i.e., active cells) in a DUNE grid view.  We form one set of `SatfuncConsistencyChecks` objects for each `SatfuncCheckPointInterface<>` object, thereby enabling running the same set of consistency checks for both the unscaled, tabulated, per-region saturation functions and the per-cell scaled saturation functions.  The latter is executed only if the run enables end-point scaling for the saturation functions.

As an aid to enabling the pertinent individual checks, we add a private factory function which considers the run's active phases and whether or not the run uses the alternative, three-point horizontal scaling method ("SCALECRS = YES").

### CODE REVIEW

**Comment by bska** [comment]
The upstream PR #5589 was merged into the master branch.  I'm marking this PR as "ready for review" and I'm running a build check.

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
jenkins build this please

File: tests/test_SatfuncConsistencyCheckManager.cpp [hunk 1]
```diff
@@ -0,0 +1,1612 @@
...
+{}
+
+-- =================================================================
+REGIONS
+
+SATNUM
+ 1 /
+
+END
+)", epsSpec)))
```
Comment (atgeirr) [inline tests/test_SatfuncConsistencyCheckManager.cpp:204]: This neat trick was a bit hard to notice, would be better to put this string generation into a separate function.

**Review by atgeirr — COMMENTED** [review]
Regarding the tests: How robust/stable is it to test the explicit formatted error output? If we decide to slightly adjust the format that will give a lot of extra work. Also, the numerical precision can perhaps change etc. Is it not an alternative to have the failed check result returned as some report object that can be tested? The formatting would then be handled and tested separately.

Other than that, only minor issues.

File: opm/simulators/utils/satfunc/SatfuncConsistencyCheckManager.hpp [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+            for (const auto& elem : elements(gv, Dune::Partitions::interior)) {
+                this->runCellChecks(getCellIndex(elem));
+            }
+
+            gv.comm().barrier();
+
+            this->collectFailures(gv.comm());
+        }
+
+        /// Whether or not any checks failed at the \c Standard level.
```
[the remark this answers is not in the document]
Comment (bska) [reply opm/simulators/utils/satfunc/SatfuncConsistencyCheckManager.hpp:143]: > Is this only returning true if there were Standard failures, and not if there were Standard _or_ Critical failures?

Yes.  I've pushed an update to rename the function to `anyFailedStandardChecks()`.

File: opm/simulators/utils/satfunc/SatfuncConsistencyCheckManager.hpp [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+        ///
+        /// In a parallel run it is only safe to call this function on the
+        /// MPI process to which the consistency check violations were
+        /// collected in a previous call to collectFailures().
+        ///
+        /// \param[in] level Report's severity level.
+        ///
+        /// \param[in] emitReportRecord Call-back function for outputting a
+        /// single record/line of a violation report.  Typically a wrapper
+        /// of \code OpmLog::info() \endcode.  It is the responsibility of
```
[the remark this answers is not in the document]
Comment (bska) [reply opm/simulators/utils/satfunc/SatfuncConsistencyCheckManager.hpp:162]: The suggestion of `OpmLog::info()` was actually intentional.  I don't really want large amounts of coloured text added to the interactive terminal if there are lots of failures.  Indeed, the downstream PR #5596 uses `OpmLog::warning()` and `OpmLog::error()` to print leading headers that call attention to the issues, but the individual problems/failures are reported using `OpmLog::info()`.

File: opm/simulators/utils/satfunc/SatfuncConsistencyCheckManager.cpp [hunk 4]
```diff
@@ -0,0 +1,506 @@
...
+SatfuncConsistencyCheckManager(const std::size_t    numSamplePoints,
+                               const EclipseState&  eclipseState,
+                               const LocalToGlobal& localToGlobal)
+    : eclipseState_  { std::cref(eclipseState) }
+    , localToGlobal_ { localToGlobal }
+    , rtep_          { rawTableEndpoints(eclipseState) }
+    , rfunc_         { rawFunctionValues(eclipseState, rtep_) }
+{
+    // Note: This setup is limited to
+    //   1. Drainage only--no hysteresis
```
[the remark this answers is not in the document]
Comment (bska) [reply opm/simulators/utils/satfunc/SatfuncConsistencyCheckManager.cpp:97]: > Is this precise?

Yes.

> There are parts referring to checking the imbibition curve further down.

True.  There are parts here that are independent of the curve type, and I very much intend to extend this to cover imbibition in the future.  However, the `configureCurveChecks()` function is presently very much specific to drainage processes without directional variation.  In order to do that, however, the set of checks&ndash;e.g., the [`OilPhaseConsistencyChecks`](https://github.com/OPM/opm-simulators/blob/c228f31287279bd7b93446efbbcb7c4ae72ef742/opm/simulators/utils/satfunc/OilPhaseConsistencyChecks.hpp)&ndash;must at least become aware of the alternate end-point names to report the proper header names.

**Review by bska — COMMENTED** [review]
> Regarding the tests: How robust/stable is it to test the explicit formatted error output?

Maybe not as much as we'd like.  However, the output format for floating point numbers is pinned down in the [`writeTableRecord‎()`](https://github.com/OPM/opm-simulators/blob/c228f31287279bd7b93446efbbcb7c4ae72ef742/opm/simulators/utils/satfunc/SatfuncConsistencyChecks.cpp#L460) helper function and I actually do want to test that it gives the expected output.

> If we decide to slightly adjust the format that will give a lot of extra work.

True.

> Also, the numerical precision can perhaps change etc

Only if we explicitly decide to change it.

> Is it not an alternative to have the failed check result returned as some report object that can be tested?

That possible of course, but it would effectively be the entire internal state of the `SatfuncConsistencyCheck` object.

> The formatting would then be handled and tested separately.

This test is supposed to be the check for the output formatting.  It is kind of an integration test for the entire component.  There are other tests that probe the behaviour, e.g., whether a failure is at the "Standard" or "Critical" levels&ndash;and numerical values reported of individual consistency checks.  See, for instance, [`tests/test_OilSatfuncConsistencyChecks.cpp`](https://github.com/OPM/opm-simulators/blob/c228f31287279bd7b93446efbbcb7c4ae72ef742/tests/test_OilSatfuncConsistencyChecks.cpp).

File: tests/test_SatfuncConsistencyCheckManager.cpp (hunk 1 above)
Comment (bska) [reply tests/test_SatfuncConsistencyCheckManager.cpp]: > This neat trick was a bit hard to notice, would be better to put this string generation into a separate function.

I've added a helper function that returns a `Deck` object instead.  I didn't see any benefit to having a function of the form
```C++
std::string makeEpsString(std::string_view epsSpec)
{
    return fmt::format(R"(RUNSPEC
...)", epsSpec);
}
```
as the return value would just be passed immediately into `Parser{}.parseString()`.

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
jenkins build this please

**Comment by atgeirr** [comment]
Thanks for explanations and revisions. Merging!

**Pull request merged**